### ISSUE TITLE
[Bug]: A5 compile UB overflow for bf16 online softmax with tl.static_range (V=151936); tl.range OK

### ISSUE THREAD

**Issue by zheliuyu** [triton-lang/triton-ascend#2246]
### Summary

On **Ascend A5** (Ascend950PR, `Ascend950PR_9579` compile target), a minimal **online softmax over a large vocabulary** kernel **fails at compile time** with **UB overflow** when the vocab scan uses `tl.static_range(0, N, BLOCK_N)`, but **compiles and runs** when the same logic uses `tl.range(0, N, BLOCK_N)`.

Typical production shape: `N = 151936` (LLM vocab), `BLOCK_N = 8192` → **19** loop iterations. The failure is in **`hivm-plan-memory` / `buildFinalHIVMPipelines`**, not a runtime numeric error.

Root cause class (isolated on A5):

1. **bf16** masked load converted to **fp32** inside the loop (as in GRPO / log-softmax kernels).
2. Inner loop written with **`tl.static_range`**, so the A5 backend **fully unrolls** the trip count and appears to **retain ~one fp32 `BLOCK_N` tile per unrolled iteration** in UB at once (~262144 bits per 8192-wide fp32 vector).
3. **`tl.range`** lowers to a **dynamic** loop; planning keeps **one** live tile → compile OK.

The **same Triton source** with `tl.static_range` is reported to compile on **Ascend A3** (A2/A3 pipeline does not account for UB this way). **`multibuffer=False`** does not reduce the reported requirement on A5 for this repro (`requires=4333568` unchanged in our tests).

This blocks downstream **GRPO / CISPO / SAPO**-style kernels that scan `V ≈ 151936` with production block sizing.

### Minimal complete example

Save as `repro_grpo_static_range_ub.py` and run on an idle A5 NPU:

```bash
python repro_grpo_static_range_ub.py
```

```python
#!/usr/bin/env python3
"""Minimal A5: tl.static_range online softmax UB vs tl.range."""

import re
import sys

import torch
import triton
import triton.language as tl

N = 151936
BLOCK_N = 8192
INV_TEMP = 1.0 / 0.7

OVERFLOW_RE = re.compile(r"requires (\d+) bits while (\d+) bits available", re.I)


def classify(exc: BaseException) -> str:
    text = str(exc)
    m = OVERFLOW_RE.search(text)
    if m:
        return f"COMPILE_UB requires={m.group(1)} avail={m.group(2)}"
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    return (lines[-1] if lines else type(exc).__name__)[:200]


def try_launch(label: str, fn) -> None:
    try:
        fn()
        torch.npu.synchronize()
        print(f"{label}: OK")
    except Exception as e:
        print(f"{label}: FAIL ({classify(e)})")


@triton.jit
def online_softmax_static(ptr, inv_temp, N: tl.constexpr, BLOCK_N: tl.constexpr):
    m_i = float("-inf")
    l_i = 0.0
    for start in tl.static_range(0, N, BLOCK_N):
        cols = start + tl.arange(0, BLOCK_N)
        x = tl.load(ptr + cols, mask=cols < N, other=float("-inf")).to(tl.float32) * inv_temp
        new_m = tl.maximum(m_i, tl.max(x))
        l_i = l_i * tl.exp(m_i - new_m) + tl.sum(tl.exp(x - new_m))
        m_i = new_m
    tl.store(ptr, (m_i + tl.log(l_i)).to(ptr.dtype.element_ty))


@triton.jit
def online_softmax_range(ptr, inv_temp, N: tl.constexpr, BLOCK_N: tl.constexpr):
    m_i = float("-inf")
    l_i = 0.0
    for start in tl.range(0, N, BLOCK_N):
        cols = start + tl.arange(0, BLOCK_N)
        x = tl.load(ptr + cols, mask=cols < N, other=float("-inf")).to(tl.float32) * inv_temp
        new_m = tl.maximum(m_i, tl.max(x))
        l_i = l_i * tl.exp(m_i - new_m) + tl.sum(tl.exp(x - new_m))
        m_i = new_m
    tl.store(ptr, (m_i + tl.log(l_i)).to(ptr.dtype.element_ty))


def main() -> int:
    torch.npu.set_device(0)
    print("device:", torch.npu.get_device_name(0))
    print(f"N={N} BLOCK_N={BLOCK_N} unrolls={(N + BLOCK_N - 1) // BLOCK_N}")
    x = torch.empty(N, device="npu", dtype=torch.bfloat16)
    x.fill_(0.01)
    grid = (1,)
    try_launch("control_tl_range", lambda: online_softmax_range[grid](x, INV_TEMP, N=N, BLOCK_N=BLOCK_N))
    try_launch("fail_static_range", lambda: online_softmax_static[grid](x, INV_TEMP, N=N, BLOCK_N=BLOCK_N))
    return 0


if __name__ == "__main__":
    sys.exit(main())
```

### Expected vs observed

**Expected**

- Both kernels compile; both compute the same log-sum-exp style reduction over `N` bf16 elements (up to masking).

**Observed (A5, Ascend950PR_9579)**

```
device: Ascend950PR_9579
N=151936 BLOCK_N=8192 unrolls=19
control_tl_range: OK
fail_static_range: FAIL (COMPILE_UB requires=4333568 avail=2031616)
```

Compiler diagnostics (abbreviated):

```
loc(".../repro_grpo_static_range_ub.py":...): error: Failed to run buildFinalHIVMPipelines pipeline
loc(".../repro_grpo_static_range_ub.py":...): error: ub overflow, requires 4333568 bits while 2031616 bits available!
[ERROR] Failed to run BiShengIR pipeline
```

Notes on the numbers:

- Available **2031616** bits ≈ **198 KiB** (typical AIV UB budget in the error).
- Required **4333568** bits ≈ **16.5×** one fp32 vector of length 8192 (**262144** bits each), consistent with full unroll of ~19 tiles (compiler saturates below 19 in practice).

Invoked backend (typical footer):

```
bishengir-compile-a5 ... --target=Ascend950PR_9579 --enable-auto-multi-buffer=True ... --enable-triton-kernel-compile=true ...
```

### Bisection matrix (A5, `BLOCK_N=8192`, default multibuffer)

| Kernel / loop | `N=65536` (8 unrolls) | `N=151936` (19 unrolls) |
|---------------|------------------------|-------------------------|
| bf16 copy, `static_range`, no fp32 convert | OK | OK |
| bf16 load + immediate `tl.sum` → scalar, `static_range` | OK | OK |
| fp32 load online softmax, `static_range` | OK | OK |
| **bf16→fp32 online softmax, `static_range`** | **COMPILE_UB ~2232064** | **COMPILE_UB 4333568** |
| Same, **`tl.range`** | **OK** | **OK** |
| Same `static_range`, **`multibuffer=False`** | (same class) | **still 4333568** |

Optional extended bisection scripts: `min_repro_grpo_root.py` (same repo experiments tree).

### Downstream impact

In [Liger-Kernel](https://github.com/linkedin/Liger-Kernel) Ascend **GRPO loss**, pytest cases with `V=151936` (e.g. `test_grpo_loss`, `test_cispo_loss`, `test_sapo_loss` with `1-1024-151936-...`) failed on A5 with the production `_grpo_loss_fwd_kernel` until inner vocab loops were switched from `tl.static_range` to `tl.range`. Block sizing assumed **one** live `BLOCK_N` tile in UB; that assumption holds on A3 but not on A5 with `static_range` + bf16→fp32.

### Workaround (application side)

Use **`tl.range(0, N, BLOCK_N)`** instead of **`tl.static_range(0, N, BLOCK_N)`** for large-`N` vocab scans. **Not** acceptable as the only story: many kernels rely on `static_range` for optimization on other backends; A5 should not force dynamic loops for correctness of UB planning. Shrinking `BLOCK_N` (e.g. 2048) can **pass compile** but **fault at runtime** with UB access overflow for large `N` — not a safe fix.

### Environment details

```
Triton: 3.2.0
triton_ascend: 3.2.2 (pip package name triton_ascend)
PyTorch: 2.9.0+cpu
torch_npu: 2.9.0.post4
NPU: Ascend950PR (reported device name: Ascend950PR_9579)
OS: Linux (openEuler 24.03 SP3, kernel 6.6.x)
Python: 3.11
```

Compile path: 
A5 — `bishengir-compile-a5 --target=Ascend950PR_9579` (with enable-auto-multi-buffer, triton-kernel-compile, hfusion, etc. as in default triton-ascend A5 pipeline)

Cross-SOC: 

Same Triton kernel with `tl.static_range` compiles on Ascend A3 (A2/A3 pipeline); compile-time UB overflow observed only on A5 pipeline above for this repro.

Repro command:
```
  python repro_grpo_static_range_ub.py
  # first JIT ~30–60s; control kernel OK, failing kernel fails at compile (no launch)
```

**Comment by hongziqi** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.
